### PR TITLE
Ai

### PR DESCRIPTION
**Pull request by hklerer**

### CODE REVIEW

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+public class GeneticLearn {
+
+    private static final int GENERATIONS = 10;
+    private static final int NETWORK_COUNT = 1000;
+    private static final int INPUT_SIZE = 2;
+    private static final int OUTPUT_SIZE = 2;
+    private static final int TOP_AMOUNT = (int) (NETWORK_COUNT * .10);
+    public static final int HIDDEN_NODES = 128;
+
+    public static void main(String[] args) {
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: How long does it take to run this?

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+        // For every generation
+        for (int i = 0; i < GENERATIONS; i++) {
+            int totalScore = 0;
+            int totalticks = 0;
+            Map<NeuralNetwork, NetworkStats> networkStatsMap = new HashMap<>();
+            // Every network plays the game
+            for (NeuralNetwork network : networks) {
+                BBController bbController = new BBController();
+                BBFrame frame = new BBFrame(bbController);
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: Do you really need a frame to do this?

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                frame.setVisible(true);
+                boolean running = bbController.isRunning();
+                while (running) {
+                    double[] input = new double[INPUT_SIZE];
+                    input[0] = bbController.currAngle;
+                    input[1] = bbController.getScore();
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: Why is the score part of the input?

**Comment by andrewoid** [comment]
Did you succeed in training the UI to follow the ball?

File: src/main/java/bfk/brickbreaker/GeneticLearn.java (hunk 1 above)
Comment (annafrei924) [reply src/main/java/bfk/brickbreaker/GeneticLearn.java]: Every generation takes about 30 minutes, is there a way to lessen that?

Comment (andrewoid) [reply src/main/java/bfk/brickbreaker/GeneticLearn.java]: Well you don't need a Frame

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    private static final int OUTPUT_SIZE = 2;
+    private static final int TOP_AMOUNT = (int) (NETWORK_COUNT * .01);
+    public static final int HIDDEN_NODES = 128;
+
+    public static void main(String[] args) {
+
+        // Create Array of Networks
+        List<NeuralNetwork> networks = new ArrayList<>();
+        for (int i = 0; i < NETWORK_COUNT; i++) {
+            networks.add(new NeuralNetwork(INPUT_SIZE, HIDDEN_NODES, OUTPUT_SIZE));
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: I'm sure you don't need 128 hidden nodes

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            Map<NeuralNetwork, NetworkStats> networkStatsMap = new HashMap<>();
+            // Every network plays the game
+            for (NeuralNetwork network : networks) {
+                System.out.println("Network: " + networkNum++);
+                int success = 0;
+                BBController bbController = new BBController();
+                bbController.reset();
+                //BBFrame frame = new BBFrame(bbController);
+                //frame.setVisible(true);
+                bbController.startTimer();
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: Why do you need a timer? That was used to update the UI

File: src/main/java/bfk/brickbreaker/GeneticLearn.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        for (int i = 0; i < NETWORK_COUNT; i++) {
+            networks.add(new NeuralNetwork(INPUT_SIZE, HIDDEN_NODES, OUTPUT_SIZE));
+        }
+
+        // For every generation
+        for (int i = 0; i < GENERATIONS; i++) {
+            int tickCounter = 0;
+            int sumSuccess = 0;
+            int networkNum = 0;
+            Map<NeuralNetwork, NetworkStats> networkStatsMap = new HashMap<>();
```
Comment (andrewoid) [inline src/main/java/bfk/brickbreaker/GeneticLearn.java]: Instead of using a Map, create a class that holds the network and stats.

**Comment by annafrei924** [comment]
Hi Professor Schwimmer, all the code seems to run. It takes the angle as input & the success is measured by how often the angle is 90 degrees, but the AI isn't learning to follow the ball. Do you have any pointers?

**Comment by andrewoid** [comment]
> Hi Professor Schwimmer, all the code seems to run. It takes the angle as input & the success is measured by how often the angle is 90 degrees, but the AI isn't learning to follow the ball. Do you have any pointers?

When can we meet today to discuss it?

**Comment by annafrei924** [comment]
Are you free before 12 by any chance? Or after 5:30?

**Comment by andrewoid** [comment]
> Are you free before 12 by any chance? Or after 5:30?

Come to the zoom now.

**Comment by annafrei924** [comment]
Hi Professor Schwimmer, The results consistently improve along the first 10 generations, but then stagnate for the remaining 40 generations. Do you know why this may be happening? No matter what value we test for MAX_ROUNDS (the most rounds any given network can do), the top 10 networks in every generation are getting that maximum value.

**Comment by andrewoid** [comment]
> Hi Professor Schwimmer, The results consistently improve along the first 10 generations, but then stagnate for the remaining 40 generations. Do you know why this may be happening? No matter what value we test for MAX_ROUNDS (the most rounds any given network can do), the top 10 networks in every generation are getting that maximum value. 

What is the maximum value?

**Comment by annafrei924** [comment]
Right now it's 1000. Tried it for values between 100-10,000

**Comment by andrewoid** [comment]
What I meant was you said that it constantly improves. How well does it improve to?

**Comment by annafrei924** [comment]
For MAX_ROUNDS 1,000, the highest values in generations 10-50 are around 650,000 ticks, or avg 650 ticks/network.

**Comment by andrewoid** [comment]
> For MAX_ROUNDS 1,000, the highest values in generations 10-50 are around 650,000 ticks, or avg 650 ticks/network. 

Ok... Now take the best performing network and have it play the game, visually so you can see it.

**Comment by annafrei924** [comment]
The ai learned to go all the way to the right and try to get the ball in a loop in which it will always hit the right corner

**Pull request merged**